src/fdc/D88DiskImage.cc

```diff
@@ -0,0 +1,350 @@
+#include "D88DiskImage.hh"
+#include "RawTrack.hh"
+#include "DiskExceptions.hh"
+#include "File.hh"
+#include "FilePool.hh"
+#include "narrow.hh"
+#include "one_of.hh"
+#include "ranges.hh"
+#include "xrange.hh"
+#include <array>
+#include <cassert>
+
+namespace openmsx {
+
+static constexpr int32_t D88_TRACK_MAX = 164;
+static constexpr uint8_t D88_DISK_TYPE_2DD = 0x10;
+static constexpr uint8_t D88_DISK_TYPE_1DD = 0x40;
+
+static constexpr uint8_t D88_STATUS_ID_CRC_ERROR = 0xA0;
+static constexpr uint8_t D88_STATUS_DATA_CRC_ERROR = 0xB0;
+static constexpr uint8_t D88_STATUS_NO_ADDRESS_MARK = 0xE0;
+static constexpr uint8_t D88_STATUS_NO_DATA_MARK = 0xF0;
+
+struct D88Header
+{
+  std::array<uint8_t, 17> diskName;
+  std::array<uint8_t, 9> reserved;
+  uint8_t writeProtected;
+  uint8_t diskType;
+  int32_t diskSize;
+};
+
+struct D88SectorHeader
+{
+	uint8_t c;
+	uint8_t h;
+	uint8_t r;
+	uint8_t n;
+	uint16_t sectorsPerTrack;
+	uint8_t densityType;
+	uint8_t deleteFlag;
+	uint8_t status;
+	std::array<uint8_t, 5> reserved;
+	uint16_t sectorDataSize;
+};
+
+[[nodiscard]] static /*constexpr*/ bool isValidD88Header(const D88Header& header, const std::array<int32_t, 164>& trackOffsets)
+{
+	if (header.writeProtected != one_of(0x00, 0x10)) {
+		return false;
+	}
+	if (header.diskType != one_of(D88_DISK_TYPE_2DD, D88_DISK_TYPE_1DD)) {
+		return false;
+	}
+
+    // check track offset
+    uint32_t prevTrackOffs = trackOffsets[0];
+	if (prevTrackOffs != sizeof(D88Header) + 4 * D88_TRACK_MAX)
+		return false;
+    for(int32_t i=1; i<160; ++i)
+    {
+        auto trackOffs = trackOffsets[i];
+        auto trackSize = trackOffs - prevTrackOffs;
+        if (trackSize <= 128 || trackSize >= 0x4000)
+            return false;
+		prevTrackOffs = trackOffs;
+    }
+    for(int32_t i=160; i<D88_TRACK_MAX; ++i)
+    {
+        if (trackOffsets[i] != 0)
+            return false;
+    }
+
+	return std::ranges::all_of(header.reserved, [](auto& r) { return r == 0; });
+}
+
+D88DiskImage::D88DiskImage(const Filename& filename, std::shared_ptr<File> file_)
+    : Disk(DiskName(std::move(filename)))
+    , file(std::move(file_))
+	, trackData(8192)
+	, cachedTrackNo(-1)
+{
+    D88Header header;
+    file->seek(0);
+    file->read(std::span{&header, 1});
+    file->read(std::span{&trackOffsets, 1});

```

Comment:
If I read the [D88 File Format documentation](https://www.pc98.org/project/doc/d88.html) correctly, then older D88 disk images may only contain 160 track offsets (not 164). Should we support those? For me it's fine to not support those to keep the code simpler.